### ISSUE TITLE
Pydantic V2 doesn't handle datetimes with nanoseconds

### ISSUE THREAD

**Issue by josecsotomorales** [pydantic/pydantic#6512]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

I'm attempting to parse a datetime with nanoseconds and it's failing with validation error, example: `2023-07-07T11:10:06.008815566Z`

Error:
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for MyModel
input
  Input should be a valid datetime, second fraction value is more than 6 digits long [type=datetime_parsing, input_value='2023-07-07T11:10:06.008815566Z', input_type=str]
    For further information visit https://errors.pydantic.dev/2.1.2/v/datetime_parsing
```

Is this because of `speedate` not supporting nanoseconds at all? Any workarounds? Thanks!

### Example Code

```Python
from pydantic import BaseModel
from datetime import datetime


class MyModel(BaseModel):
    input: datetime


MyModel.model_validate({'input': '2023-07-07T11:10:06.008815566Z'})
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: /Users/Jose/Workspace/test/venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.4 (main, Jun  7 2023, 00:34:59) [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-13.4.1-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @hramezani

**Comment by Lisser** [comment]
I also encountered this. I made a (pretty ugly IMO) workaround using a validator.

```python
import re

from pydantic import BaseModel, field_validator
from datetime import datetime


class MyModel(BaseModel):
    input: datetime

    @field_validator("input", mode="before")
    def strip_nanosedonds(cls, v):
        if isinstance(v, str):
            # remove any digits after first 6 after the dot
            return re.sub(r"(\.\d{6})\d+", r"\g<1>", v)
        return v


MyModel.model_validate({'input': '2023-07-07T11:10:06.008815566Z'})
```

**Comment by adriangb** [comment]
So the real problem is that `datetime.datetime` doesn't support anything beyond microseconds anyway. That is, even if we supported parsing of arbitrary precision timestamps in our rust / pydantic-core side we have nowhere to put that information when converting to python types (unless we used a type other than `datetime` but I don't think that's a good solution).

What we are planning on doing (a process I'll kick off today) is to match the standard library's behavior when parsing datetime objects from strings: https://docs.python.org/3/library/datetime.html#datetime.time.fromisoformat. They just truncate the seconds to 6 digits.
As far as I know no one complained about this behavior in Pydantic v1.

**Comment by josecsotomorales** [comment]
The proposed solution will work, I'm already truncating it on the data producer service to workaround this error anyways, thanks!

**Comment by RS-Credentive** [comment]
I hit this problem, and I want to thank you all for addressing it!

**Comment by kjwpaulson** [comment]
> As far as I know no one complained about this behavior in Pydantic v1.

For reference the behaviour in Pydantic v1 was apparently to truncate:
```python
from pydantic import BaseModel, utils
from datetime import datetime


class MyModel(BaseModel):
    input: datetime

in_dict = {'input': '2023-07-07T11:10:06.008815966Z'}
test = MyModel(**in_dict)
print(utils.version_info())
print(f"{test.input=}")
"""
             pydantic version: 1.10.11
            pydantic compiled: True
                 install path: ..../.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.2 (main, May 30 2023, 17:45:26) [GCC 12.2.0]
                     platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.37
     optional deps. installed: ['email-validator', 'typing-extensions']
test.input=datetime.datetime(2023, 7, 7, 11, 10, 6, 8815, tzinfo=datetime.timezone.utc)
"""
```

Here is v2 behaviour:
```python
from pydantic import BaseModel, version
from datetime import datetime


class MyModel(BaseModel):
    input: datetime

print(version.version_info())
in_dict = {'input': '2023-07-07T11:10:06.008815966Z'}
test = MyModel(**in_dict)
print(f"{test.input=}")
"""
             pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: ..../.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.2 (main, May 30 2023, 17:45:26) [GCC 12.2.0]
                     platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.37
     optional deps. installed: ['typing-extensions']
Traceback (most recent call last):
  File "..../test.py", line 10, in <module>
    test = MyModel(**in_dict)
           ^^^^^^^^^^^^^^^^^^
  File "..../.venv/lib/python3.11/site-packages/pydantic/main.py", line 150, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 1 validation error for MyModel
input
  Input should be a valid datetime, second fraction value is more than 6 digits long [type=datetime_parsing, input_value='2023-07-07T11:10:06.008815966Z', input_type=str]
    For further information visit https://errors.pydantic.dev/2.1.2/v/datetime_parsing
"""
```

**Comment by guillp** [comment]
Also encountered this. My workaround is to use python-datetutils to parse the high-precision datetimes, in an [annotated custom type](https://docs.pydantic.dev/2.0/usage/types/custom/#composing-types-via-annotated):

```python 
from typing import Annotated
from datetime import datetime

from dateutil.parser import isoparse
from pydantic import BaseModel, BeforeValidator

PreciseDatetime = Annotated[datetime, BeforeValidator(lambda x: isoparse(x))]

class MyModel(BaseModel):
    my_date: PreciseDatetime

print(MyModel.model_validate({"my_date": "2023-07-10T07:41:31.218360500Z"}).my_date)
# 2023-07-10 07:41:31.218360+00:00
```

But simply dropping the extra precision digits and parsing whatever is left, as default pydantic behaviour, makes total sense !

**Comment by RS-Credentive** [comment]
Yeah, I considered taking the datetime string (which in my data is required to be structured according to RFC 3339) and running it through datetime.fromisoformat(str).toisoformat(). This may have caught a lot of other kinds of issues and been somewhat more flexible, but the re-based approach is probably a lot faster!

**Comment by samuelcolvin** [comment]
[speedate](https://github.com/pydantic/speedate) should match RFC3339, pretty closely (ignoring this mistake by me :raised_hand_with_fingers_splayed:). If you see more errors, please let us know.

**Comment by samuelcolvin** [comment]
fixed in #762
